File: rust/ex3/src/main.rs
/*
 * Yes! You solved it! A literal like "Alex" is of type 
 * &str and the struct Person was expecting a String object,
 * so you used "Alex".to_string() instead! Also congrats 
 * on fixing the meet() method... just adding a simple 
 * *---------------------------*
 * | println!("{}", self.name);|
 * *---------------------------*
 * did the trick... :)
 *
 * Let's complicate stuff... a lot... Generics...
 * Say you have a struct Person, and a method eat() for 
 * person, that takes a food object and calls it's ate()
 * method. To add some complexity we add many types of foods...
 * each implementing their own ate() method (apples and bread)
 * But all these new foods implement this one ate() method...
 * Having this in common we've also made an interface (or how
 * rust calls it - "trait") called Food that has one method (ate()).
 * This makes it easier to define the generics of the eat method.
 *
 * Try to understand the code below... And expand it... Add new foods.
 *
 * Hint: <T: Food> means that the function expects a generic type
 * T that implements the trait (or interface) Food.
 * */


trait Food {
    fn ate(&self);
}

struct Apple;
struct Bread;

impl Food for Apple {
    fn ate(&self) {
        println!("Ate an apple!");
    }
}

impl Food for Bread {
    fn ate(&self) {
        println!("Ate some bread!");
    }
}

struct Person;

impl Person {
    pub fn eat<T: Food>(&self, food: T) {
        food.ate();
    }
}

fn main() {
    let alex = Person;
    let apple = Apple;
    let bread = Bread;

    alex.eat(apple);
    alex.eat(bread);
}

File: go/ex3/main.go
/*
 * Yes! You solved it! uint64 can't hold a string like "17", so you
 * either changed the field to a string or fixed the literal to
 * Age: 17. Also congrats on fixing Meet()... just adding a simple
 * *--------------------------------------------------*
 * | fmt.Printf("Hi! My name is %s!\n", p.Name)        |
 * *--------------------------------------------------*
 * did the trick... :)
 *
 * Let's complicate stuff... a lot... Generics...
 * Say you have a struct Person, and a method Eat() for Person,
 * that takes a food value and calls its Ate() method. To add some
 * complexity we add many types of foods... each implementing their
 * own Ate() method (apples and bread). But all these new foods
 * implement this one Ate() method... Having this in common we've
 * also made an interface called Food that has one method (Ate()).
 * This makes it easier to define the generics of the Eat method.
 *
 * Try to understand the code below... it won't compile as-is.
 *
 * Hint: unlike Rust (or Java/C++), Go methods are NOT allowed to
 * have their own type parameters -- only free-standing functions
 * can be generic. [T Food] on a method is illegal. You'll need to
 * restructure Eat so the generic type parameter lives on a function
 * instead of a method, while still calling it as naturally as
 * possible from main(). Once that compiles, add a new food (e.g.
 * Pizza) and make Person eat it too.
 * */

package main

import "fmt"

type Food interface {
	Ate()
}

type Apple struct{}
type Bread struct{}

func (Apple) Ate() { fmt.Println("Ate an apple!") }
func (Bread) Ate() { fmt.Println("Ate some bread!") }

type Person struct{}

func (p Person) Eat[T Food](food T) {
	food.Ate()
}

func main() {
	alex := Person{}
	apple := Apple{}
	bread := Bread{}

	alex.Eat(apple)
	alex.Eat(bread)
}
